text/006-bft-based-ordering-service.md

```diff
@@ -0,0 +1,404 @@
+
+![alt text](https://raw.githubusercontent.com/SmartBFT-Go/SmartBFT-Go.github.io/master/flow.png)
+
+
+In the `pre-prepare` phase, the leader batches requests to assemble a proposal, and it then assigns a sequence number to the proposal. Next the leader sends a `pre-prepare` message to all followers containing the proposal, its sequence number and the current view number `v`.
+
+A follower accepts the `pre-prepare` message if the proposal is valid, it is in view `v`, it is expecting this sequence number, and it has not accepted a different `pre-prepare` message for this view and sequence number.
+Once a follower accepts a `pre-prepare` message it enters the `prepare` phase by broadcasting a `prepare` message.
+The `prepare` message contains a digest of the accepted proposal, the sequence number, and `v`.
+
+Next each node waits for a quorum of `prepare` messages with the same digest, view `v`, and sequence number. 
+The quorum size should be big enough to ensure that in the intersection of any two quorums there is at least one correct node. So for example, if `n=3f+1` then the optimal quorum size is `2f+1`. 
+
+After the node is "prepared" it sends a `commit` message with the same content as the `prepare` message, as well as a signature on the proposal.
+Each node collects a quorum of `commit` messages and their attached signatures. 
+These signatures are later used to notarize the proposal, essentially creating a checkpoint. 
+Upon receiving a quorum of validated `commit` messages, the node delivers its proposal and the signatures to the application layer for storage and processing.
+
+The quorum size is calculated in the following manner: Let `n` be the total number of nodes in the algorithm and denote `f = ((n - 1) / 3)`. Then, the quorum is: `Ceil((n + f + 1) / 2)` where `Ceil` is the ceiling function. In layman's terms, the quorum is the minimal set of nodes such that every two such sets intersect in at least 1 honest node.  
+
+**View change**:
+The view change protocol is inspired by the _Synchronization Phase_ in this [BFT-SMaRt paper](http://www.di.fc.ul.pt/~bessani/publications/edcc12-modsmart.pdf) (illustrated in the attached figure taken from the BFT-SMaRt paper).
+The view change protocol is comprised of three messages: `ViewChange`, `ViewData`, and `NewView` (`Stop`, `StopData`, and `Sync` in BFT-SMaRt).
+
+![alt text](https://raw.githubusercontent.com/wiki/SmartBFT-Go/consensus/images/viewChange.png)
+
+The `ViewChange` message (the first message) is sent to all nodes by some node who suspects the leader is faulty. This message is very lightweight and includes only the next view sequence number. 
+Once collecting `2f+1` `ViewChange` messages, a non faulty node is convinced that enough nodes consider the leader as faulty, and it will send a `ViewData` message to the next view's leader (in practice, it's the view number modulo the total number of nodes).
```

Comment:
It says: 

> and it will send a `ViewData` message to the next view's leader (in practice, it's the view number modulo the total number of nodes).


> Won't all the honest nodes potentially select a different "next view's leader" at the same time? Which one 'wins'?

So, it's the node with the identifier that is the result of the (next) view number % the number of nodes. 

If there are some nodes in view `v` and some nodes in view `v'` that is different than `v` then either it's possible to collect enough votes for the next leader (be it `v` or `v'`) or it's impossible. 

If it's impossible, the nodes increase their timeout and view number and try again. Eventually, the timeouts are big enough such that they reach an agreement.

If it's possible, then a byzantine majority moves to the next view and the new leader takes over. The nodes that tried to view change to the incorrect view occasionally "probe" the rest of the nodes, and this is a chance for them to learn about the new successful view change.